### PR TITLE
Modify the challenge leaderboard page

### PR DESCRIPTION
**Pull request by Sanji515**
#### Changes proposed in this pull request:

- Modified the challenge leadeboard page, changed the url by appending the phase split id and by default setted the top phase split id in the url

- Link to live demo: http://pr-154-evalai.surge.sh

**<p align="center">Screenshot 1:</p>**

Screenshot from 2019-06-21 23-44-27

**<p align="center">Screenshot 2:</p>**

Screenshot from 2019-06-21 23-44-52

### CODE REVIEW

**Comment by RishabhJain2018** [comment]
@Sanji515 Please resolve the conflicts here.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html [hunk 1]
```diff
@@ -1,35 +1,119 @@
...
+                <mat-chip-list>
+                  <mat-chip>B</mat-chip> - Baseline submission
+                </mat-chip-list>
+              </div>
+
+              <table *ngIf="leaderboard.length > 0 && selectedPhaseSplit" class="centered highlight">
+                <thead>
+                  <tr class="content">
+                    <td data-field="rank" class="align-center">
+                      <a (click)="sortColumn!='rank' && reverseSort = false;sortColumn == 'rank' &&
```
Comment (lunayach) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html]: Push this logic to `.ts` and create a function to take care of it. (Similarly at other places in this file.)

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 2]
```diff
@@ -98,6 +113,15 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
    */
   entryHighlighted: any = null;
 
+  /**
+   * Challenge phase visibility
+   */
+  challengePhaseVisibility = {
+    owner_and_host: 1,
+    host: 2,
+    public: 3,
```
Comment (lunayach) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:127]: I see only `public` property used below in `filterPhases( )`.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 3]
```diff
@@ -319,7 +321,9 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
     const SELF = this;
     this.apiService.getUrl(API_PATH).subscribe(
       data => {
+        console.log(data['results']);
```
Comment (lunayach) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: Remove this.

File: src/app/components/challenge/challengesubmit/challengesubmit.component.html [hunk 4]
```diff
@@ -58,249 +58,250 @@
             <button class="btn ev-btn-dark btn-waves-effect waves-dark grad-btn
               code-btn" value="evalai push <image>:<tag> --phase <phase_name>"
               (click)="copyTextToClipboard(evalai_push)">
-              <i class="fa fa-clipboard" aria-hidden="true" alt="Copy to
-                    clipboard."></i>
+              <i class="fa fa-clipboard" aria-hidden="true" alt="Copy to clipboard.">
```
Comment (lunayach) [inline src/app/components/challenge/challengesubmit/challengesubmit.component.html:61]: Why do we need to change this file in `leaderboard page` PR? Also, only changes of re-ordering the code seem to be there. @Sanji515 Can you confirm?

File: src/app/components/utility/selectphase/selectphase.component.ts [hunk 5]
```diff
@@ -76,6 +101,20 @@ export class SelectphaseComponent implements OnInit, OnChanges {
     this.phaseSelected(phase);
   }
 
+  /**
+   * Select a particular phase split.
+   * @param phase  phase split to be selected.
+   */
```
Comment (lunayach) [inline src/app/components/utility/selectphase/selectphase.component.ts:109]: Update `@param`'s above.

File: src/styles/base.scss [hunk 6]
```diff
@@ -328,7 +334,33 @@ mat-chip {
...
+.update-page {
+  display: block;
+  background: $dark-gray;
+  width: 100%;
+  overflow: auto;
+  padding: 10px;
+  text-align: center;
+  color: #fff;
+  position: absolute;
+  width: auto;
```
Comment (lunayach) [inline src/styles/base.scss:348]: If this, then remove `width: 100%;` above.

File: src/styles/base.scss [hunk 7]
```diff
@@ -328,7 +334,33 @@ mat-chip {
...
+  right: 0;
+  padding: 10px 30px;
```
Comment (lunayach) [inline src/styles/base.scss:350]: Same, remove the `padding` defined above.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 2 above)
Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:127]: @lunayach I think it is better to at least list all the available options as they are same on the backend side

File: src/app/components/challenge/challengesubmit/challengesubmit.component.html (hunk 4 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesubmit/challengesubmit.component.html:61]: @lunayach There are not much changes here but formatted the code

File: src/app/components/challenge/challengesubmit/challengesubmit.component.html [hunk 8]
```diff
@@ -58,249 +58,250 @@
...
+        Please fill the following fields to submit
+      </strong>
+    </div>
+    <div class="row">
+      <div class="col-md-7 col-sm-7 col-xs-10">
+        <div class="exist-team-card">
+          <ul>
+            <b class="fw-bold content fs-14">Select phase:</b>
+            <app-selectphase [phases]="filteredPhases"
+              [phaseSelectionListType]="phaseSelectionListType"
```
Comment (Sanji515) [inline src/app/components/challenge/challengesubmit/challengesubmit.component.html:159]: @lunayach Only this is the change that is new in this file

**Comment by lunayach** [comment]
@RishabhJain2018 it works fine except when we click on `Page is Outdated, Click to Update` button , `Last Submission at` disappears. I discussed, this with @Sanji515, he said this worked fine before that Worker issue in `ngx` when he had created it. Now, he cannot test this because the Worker is failing.

**Comment by RishabhJain2018** [comment]
I've opened the PR for the worker fix. @Sanji515 Please take a look at that and then test it.

File: src/styles/base.scss (hunk 7 above)
Comment (Sanji515) [reply src/styles/base.scss:350]: done

File: src/styles/base.scss (hunk 6 above)
Comment (Sanji515) [reply src/styles/base.scss:348]: done

File: src/app/components/utility/selectphase/selectphase.component.ts (hunk 5 above)
Comment (Sanji515) [reply src/app/components/utility/selectphase/selectphase.component.ts]: done

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts (hunk 3 above)
Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts]: done

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html (hunk 1 above)
Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.html]: @lunayach done

**Comment by Sanji515** [comment]
@RishabhJain2018 @lunayach Resolved all the comments and tested the functionality, it's working perfectly on my side. Please review :smiley:

File: src/app/components/utility/selectphase/selectphase.component.ts (hunk 5 above)
Comment (Sanji515) [reply src/app/components/utility/selectphase/selectphase.component.ts:109]: done

**Comment by lunayach** [comment]
Hi @Sanji515, I was referring to this issue. When I click on the update button, the `Last Submission at` disappears. (Using EvalAI's backend while doing `ng serve`.)
record

**Comment by RishabhJain2018** [comment]
@Sanji515 The issue mentioned by Mayank is fixed, right?

**Comment by lunayach** [comment]
> @Sanji515 The issue mentioned by Mayank is fixed, right?

The issue is resolved now. Thanks, @Sanji515.

File: src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts [hunk 9]
```diff
@@ -221,18 +236,18 @@ export class ChallengeleaderboardComponent implements OnInit, AfterViewInit {
...
-      } else if (SELF.router.url.indexOf(phaseSplit['phase_split']['id']) < 0 && SELF.router.url.split('/').length === 5) {
-        SELF.router.navigate(['../' + phaseSplit['phase_split']['id']], {relativeTo: this.route});
-      } else if (SELF.router.url.indexOf(phaseSplit['phase_split']['id']) < 0 && SELF.router.url.split('/').length === 6) {
-        SELF.router.navigate(['../../' + phaseSplit['phase_split']['id']], {relativeTo: this.route});
-      } else {
-        SELF.selectedPhaseSplit = phaseSplit;
-        if (SELF.selectedPhaseSplit['phase_split']) {
-          SELF.fetchLeaderboard(SELF.selectedPhaseSplit['phase_split']['id']);
-        }
+        SELF.router.navigate([phaseSplit['id']], {relativeTo: this.route});
```
Comment (lunayach) [inline src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:239]: @Sanji515 can you also refer this file in #171?

Comment (Sanji515) [reply src/app/components/challenge/challengeleaderboard/challengeleaderboard.component.ts:239]: sure, thanks :+1:

**Review by lunayach — APPROVED** [review]
This looks good to be merged now.
Two points:

- Currently, the "**Page is Outdated, Click to update**" button appears only when the no. of submissions for that particular phase and split has changed. But what if the host re-runs the submission of the user(s). I think an additional check for this should also be included.
- The toggle button, **Show on Leaderboard** does not seem to be rendering properly. 
toggle

@Sanji515 @RishabhJain2018 should we create issues to take care of the above points later. (IDK, maybe the second point could be addressed in this PR itself.)

**Comment by RishabhJain2018** [comment]
@Sanji515 Can you please take care of both the things in a separate PR?

**Pull request merged**